fix(dev): replace config.define instead of inserting object

sapphi-red opened the pull request:
When `config.define` is set like below (`vite.config.js`)
```js
module.exports = {
  define: {
    __VAR__: 'string'
  }
}
```
and `main.js` is like below.
```js
console.log(__VAR__)
```

On dev server it becomes like
```js
console.log('string')
```
though built script becomes like
```js
console.log(string)
```

This PR will fix the build to be same as dev server.

antfu commented:
This is the expected behavior. Users should wrap it with `JSON.stringify` explicitly if they expected it is a string. Or you may need to write something like `console.log("__VAR__")` in your source code. We should update the docs to proper documentation though. Thanks for your contribution but sorry as I am closing this for now.

sapphi-red commented:
So you mean the dev server's output(`console.log('string')`) is a bug and it should be `console.log(string)`?
Or is it expected that the output differs between dev server and build?

antfu commented:
Sorry, it's my bad, I misread the issue, could you reopen it? Thanks 😅

antfu commented:
And yes, the bug is on the dev side.

sapphi-red commented:
I have restored the branch and made a change to fix it.
I am not sure whether this works properly, but it worked when I wrote `__VAR__` inside `.ts`.

sapphi-red commented:
I am sorry I forgot checking the test.
I will resolve conflicts and fix it.

sapphi-red commented:
I have fixed it and added a test! 😃

antfu reviewed (approved):
LGTM. @underfin I would need some double review to be sure, thanks.

antfu commented on src/node/server/serverPluginModuleRewrite.ts:
```diff
@@ -137,13 +146,17 @@ export function rewriteImports(
       )
     }
 
+    const hasDefined = Object.keys(replacements).some((replacement) =>
```
`hasDefined` is a bit unclear here, not sure about the proper naming here, maybe `needReplacements`?

sapphi-red replied:
Thanks, that name is more clear.

underfin commented:
It's better if move replace logic into `serverReplacePlugin`, because it is not related of `module` rewrite.

sapphi-red commented:
I have moved it into `serverPluginReplace.ts` and changed it not to replace inside `<template>` for consistency.
https://github.com/vitejs/vite/blob/47eb448709fa09f8bd02e1d4159d68e467ed68a7/src/node/build/index.ts#L352-L361

underfin committed:
fix: fix code style

underfin commented:
I rethink the orignal issues and we should fix build.

sapphi-red commented:
Why did you think so? (I just want to know the reason.)
Is it because replacing to a raw source code is a bit curious?

underfin commented:
It is unnecessary and has  some overhead of transfromed file can be replaced.It can be apply correct value by `window`, we should follow the pervious implement.

sapphi-red commented:
I see. Thanks for answering.
